Prefer failures

notatallshaw opened the pull request:
PR for https://github.com/pypa/pip/issues/10479

I actually think it would make sense to prefer *all* the ancestors of the failures, but I couldn't figure out a simple way to do that.

notatallshaw commented:
> I actually think it would make sense to prefer _all_ the ancestors of the failures, but I couldn't figure out a simple way to do that.

After a bit of thought I realized that older ancestors are all going to be pinned so preferring all ancestors would probably at best only help situations that required deep backtracking.

I did manage to implement a relatively simple version thought and it worked well with all my known test cases, but didn't seem to provide any additional benefit. Further it seems like it isn't strong enough to tackle the Airflow 1.10.13 case. So I think trying to prefer all the ancestor adds a bit of complexity for no provable benefit. Can revisit in the future though.

uranusjr commented:
> I did manage to implement a relatively simple version and it worked well with all my known test cases, but didn't seem to provide any additional benefit.

I think I may not be reading this correctly, but what are the additional benefits you feel should be expected? The patch only modifies the backtracking behaviour from I can tell (because `failure_causes` is always empty outside the backtracking context), so it makes sense to me it does not provide any benefits outside its intended target.

notatallshaw commented:
> 
> 
> > I did manage to implement a relatively simple version and it worked well with all my known test cases, but didn't seem to provide any additional benefit.
> 
> I think I may not be reading this correctly, but what are the additional benefits you feel should be expected? The patch only modifies the backtracking behaviour from I can tell (because `failure_causes` is always empty outside the backtracking context), so it makes sense to me it does not provide any benefits outside its intended target.

You are correct that this patch only modifies the backtracking behavior. Sorry I did not explain myself very well, let me attempt to do better:

Say you had 3 requirements "X, Y, Z" and X depended on A which depended on B, and Y depended on C which depends on B. But the latest version of C and the latest version of A's dependency on B is incompatible. This creates a failure in attempting to pin at some point and causes Pip to backtrack, 

In this scenario the patch I provide is supposed to prefer up to B, C, and A. As A and C disagree about what version of B they should have, the assumption is resolving this conflict before anything else will speed up total backtracking time.

What I thought (and crossed out in my original comment now) is it might make sense in some circumstances to prefer the whole ancestor chain, so the combination of X, A, B, and Y, C, B. My reasoning for this in some circumstances it could be that X and A as well as Y and C are tightly coupled and therefore preferring A and C may not be helpful as as you need to backtrack to earlier versions of X or Y.

I created a new patch which implemented this "prefer all ancestors of failure causes" but after experimenting with it I could not find an example where it actually helped. I wanted to communicate this negative result to be clear where my thoughts on this optimization are and what I've tried.

Hope that explanation makes more sense? Seems like these ideas are somewhat difficult to communicate textually, but I am happy to answer any questions you have.

pradyunsg commented on src/pip/_internal/resolution/resolvelib/provider.py:
```diff
@@ -132,11 +133,25 @@ def get_preference(
             delay_this,
             not direct,
             not pinned,
+            not responsible_for_failure,
```
Hmm... This should likely be the first item in this tuple. We want to actively try variants of items that are involved in a conflict.

notatallshaw replied:
You are likely correct. My reasoning was python, direct, and pinned should all be singular package requirements, and I didn't want to mess with pinning them first. But as "responsible_for_failures" only becomes relevant once there are failures I guess this makes sense. I will do some sanity testing though that this doesn't cause problems.

As for "delay_this" I actually personally feel that this setuptools hack can be removed with this change, as it won't be responsible for failures and therefore pip won't be "accidentally" backtracking on setuptools anymore. I just wasn't brave enough to do it on this PR.

uranusjr replied:
We can have a `--use-deprecated=setuptools-resolve-hack` that enables `delay_this` and point people report to a tracking issue when they use this so we can get an idea if it’s actually useful after this. But I agree this should be a separate PR.

Meanwhile, I would put `not responsible_for_failure` after `delay_this` (`requires_python` should always be the first one to make unresolvable Python version conflits bubble up as soon as possible).

notatallshaw replied:
I am running through my known test cases and I found a requirements file where the proposed change here makes it significantly slower to resolve (in fact it hasn't finished resolving):

```
# Python 3.7 only
poetry2setup
wheel
twine
```

Without the proposed change it significantly backtracks on "keyring" to "19.2.0", then after 1 more significant backtrack on "importlib_metadata" it resolves quickly.

With the proposed change the resolver backtracks  on "keyring" to "15.1.0", backtracks significantly on "importlib_metadata" (but much slower), then backtracks significantly on "packaging", and that is as far as I've been able to get after an hour of backtracking.

I have not yet been able to figure out all the specific details. But I hypothesise it is very efficient to quickly check single requirements (e.g. direct and pinned), even versus following backtrack failures.

Let me know what you think.

pradyunsg commented on src/pip/_internal/resolution/resolvelib/provider.py:
```diff
@@ -132,11 +133,25 @@ def get_preference(
+        # Prefer the causes of failures. In general this speeds up backtracking,
+        # on the assumption that the problem resolving the dependency tree is
+        # related to the failures that are currently happening
+        responsible_for_failure = False
+        for failure_cause in failure_causes:
+            if identifier == failure_cause.requirement.name:
+                responsible_for_failure = True
+                break
+            if failure_cause.parent and identifier == failure_cause.parent.name:
+                responsible_for_failure = True
+                break
```
Let's have this in a function, to use the `return` statement to break out of the loop and get the value.

We can also "flip" the return values, and directly use this without needing a `not` in the tuple.

```py
def unrelated_to_conflict(identifier: str, failure_causes):
    for failure_cause in failure_causes:
        if identifier == failure_cause.requirement.name:
            return False
        if failure_cause.parent and identifier == failure_cause.parent.name:
            return False
    return True
```

notatallshaw replied:
I'm happy to do this, though I personally dislike a function or variable name starting with a negative word like "unrelated".

pradyunsg replied:
The important bit from my perspective is making it into a function to avoid the breaks -- I'm ambivalent on what names we end up with. :)

uranusjr replied:
+1 to extracting this into a function. Do what you feel best regarding the name.

notatallshaw replied:
Moved to function "is_backtrack_cause"

pradyunsg reviewed (changes requested):
I really like this -- this is a very smart way to solve the problem of actively backtracking on the items that are implicated in the conflicts.

A couple of minor comments.

uranusjr commented:
(Also, please fix the linter errors first so the actual tests can run.)

notatallshaw commented:
Added requested function and changed name to "backtrack_causes" instead of "failure_causes", I think this makes more sense  in context, as these aren't failures of the current state but of the state they we backtracked from.

notatallshaw commented:
I'm realizing this is going to require modifying a lot of test cases or at least test set-up. In most cases though it should just require adding: `backtrack_causes=[]` on to calling `get_preference`.

Could also make the `backtrack_causes` argument optional, but that feels like it could lead to accidental bugs.

But before proceeding either way I'll wait to see if you think the approach is correct, maybe there's a way I should do this without adding an extra argument.

uranusjr commented:
I think adding a new (required, not optional) argument is the way to go. Not entirely sure `backtrack_causes` is the best name though, let’s not change the tests until we land the ResolveLib changes first (which would include deciding the argument name to use since ResolveLib always call provider methods with named parameters).

pradyunsg commented on news/10479.feature.rst:
```diff
@@ -0,0 +1 @@
+New resolver: When backtracking prefer the dependencies which caused the most recent backtracking. In some cases this will significantly reduce the amount of backtracking required.
```
```suggestion
New resolver: When backtracking prefer the dependencies which are involved in the most recent conflict. In some cases this will significantly reduce the amount of backtracking required.
```

notatallshaw replied:
I have no opinion on this, but FYI I used "New resolver" as that is what was all the resolver news as recently as 21.2.2: https://github.com/pypa/pip/blob/main/NEWS.rst#bug-fixes-2

And my other PR got merged with "New resolver": https://github.com/pypa/pip/pull/10482/files

Happy to change though.

pradyunsg commented:
Not gonna run the CI, since we know it's gonna fail until we update the vendored resolvelib, which needs a resolvelib release. :)

pradyunsg commented:
FWIW, @notatallshaw -- would it be possible for you to share the list of small requirements.txt files that trigger the backtracking behaviour, so that others can see how this behaves?

notatallshaw commented:
> 
> 
> FWIW, @notatallshaw -- would it be possible for you to share the list of small requirements.txt files that trigger the backtracking behaviour, so that others can see how this behaves?

Yes, been meaning to do this for a while but it ends up being pretty time consuming to go back and recheck everything.

Where I've written Linux or Windows only it's usually because system dependencies are also required. I've also made note of where the requirements file end up in ResolutionImpossible as I found it interesting even in these cases this optimization is still highly effective at reducing backtracking.

| Requirements File | Notes | Reported Location |
| --- | --- | --- |
| [requirments_1.txt](https://github.com/pypa/pip/files/7231525/requirments_1.txt) | | https://github.com/pypa/pip/issues/10201 |
| [requirments_2.txt](https://github.com/pypa/pip/files/7231527/requirments_2.txt) |  https://github.com/pypa/pip/issues/10201 |
| [requirments_3.txt](https://github.com/pypa/pip/files/7231529/requirments_3.txt) | https://github.com/pypa/pip/issues/10373 |
| [requirments_4.txt](https://github.com/pypa/pip/files/7231530/requirments_4.txt) | Python 3.7 only |  |
| [requirements_5.txt](https://github.com/pypa/pip/files/7231536/requirements_5.txt)  | Linux only | https://github.com/pypa/pip/issues/9187#issuecomment-908961314 |
| [requirements_6.txt](https://github.com/pypa/pip/files/7231538/requirements_6.txt) | Windows Only, I personally haven' reproduced | https://github.com/pypa/pip/issues/10479#issuecomment-922333239 |
| [requirments_7.txt](https://github.com/pypa/pip/files/7231575/requirments_7.txt) | ResolutionImpossible | https://github.com/neurolibre/neurolibre/issues/13 |
| [requirments_8.txt](https://github.com/pypa/pip/files/7231582/requirments_8.txt) | | https://github.com/pypa/pip/issues/10230 |
| [requirments_9.txt](https://github.com/pypa/pip/files/7231584/requirments_9.txt) | ResolutionImpossible | https://github.com/pypa/pip/issues/9187#issuecomment-756652530 |
| [requirments_10.txt](https://github.com/pypa/pip/files/7231602/requirments_10.txt) | ResolutionImpossible, Linux Only, Must install Six first | https://github.com/pypa/pip/issues/9187 |

notatallshaw commented:
Question on Pip process: Once there is a resolvelib release who is responsible for vendoring the new version? Is is a separate PR? Do I include the updated config in this PR and then it automatically happens in the CI pipeline? Or do I update the config locally and run the vendoring sync locally and include the resolvelib changes in this PR?

pfmoore commented:
Vendoring a new release of resolvelib (or indeed any vendored package) should be done as a separate PR. There's developer docs on how to do it, I believe, but basically it's a matter of updating `vendor.txt` and running `tox -e vendoring`.

notatallshaw commented:
> 
> 
> Vendoring a new release of resolvelib (or indeed any vendored package) should be done as a separate PR. There's developer docs on how to do it, I believe, but basically it's a matter of updating `vendor.txt` and running `tox -e vendoring`.

Thanks I wasn't sure because a lot of tests will break with a new version of resolvelib until this PR lands due to changes in the `get_preference` method.

pfmoore commented:
Ah, I see. I still think it should be a separate PR, but merging the two PRs should probably be co-ordinated, so I'll leave that to @pradyunsg or @uranusjr, as they've been the most directly involved with this work.

uranusjr commented:
It's easiest to do it in this PR. We should ge the current code kind-of working (I think it's close, just need confirmation from @pradyunsg), do one commit to update resolvelib, and then merge.

edmorley commented:
Hi! Many thanks for everyone's hard work here :-) I'm hoping that this PR makes it to the 21.3 release - and from reading the above, I'm presuming that's the intention? If so, should this PR be added to the 21.3 milestone?

pradyunsg commented:
I'll add it in. I'm looking at our set of vendored dependencies right now, and will likely end up picking this up this weekend as well.

notatallshaw commented:
Once `resolvelib` lands I'll need to rebase and fix this test I recently added in a related PR: https://github.com/pypa/pip/blob/main/tests/unit/resolution_resolvelib/test_provider.py#L29

But then I think the PR will be good to go pending some more thoughts and suggestions from the PIP maintainers. I'll also probably re-run through all the test cases I can still reproduce.

notatallshaw commented:
>     * Rebase on the current `main` branch, while squashing all the commits currently in this PR into a single commit -- this should be doable with a single `git rebase -i` command.

Not sure I get this. I tried `git rebase -i` and just ended up in infinite loops of git complaining there were conflicts that I'd fix and try to commit and it'd fail or it wouldn't and then I'd just get the same conflict again.

I'm traveling right now and attending a wedding  so I don't have a lot of time to spend debugging this. Would it be easier to make a new branch and a new PR?

notatallshaw commented:
>     * Rebase on the current `main` branch, while squashing all the commits currently in this PR into a single commit -- this should be doable with a single `git rebase -i` command.


I tried doing this but it seemed to add even more commits on to this pull request. Not exactly sure what the goal is here, my knowledge of using git in such a large project is very limited.

pradyunsg commented:
Alrighty. I'm done.

notatallshaw commented:
> Would it be fine if I did squashing instead then?

Yes, absolutely, I'm on a flight right now to a friend's wedding and may not be able to spend much time over the next few days, so feel free to push whatever needs pushing.

I don't have any ownership feelings over this PR I just want to help Pip. But if you do need me to do something I'll catch up as soon as I can.

pradyunsg commented:
Thanks for the review @uranusjr! And, @notatallshaw, for all the work investigating and implementing this! ^>^

notatallshaw commented:
> Thanks for the review @uranusjr! And, @notatallshaw, for all the work investigating and implementing this! ^>^

Here's to hoping it doesn't cause more issues than it solves!

Feel free to ping me on 21.3+ heavy backtracking reports. I am very invented at this point to keep resolving this issue
